src/Engine/Renderer/Mesh/Mesh.hpp

```diff
@@ -79,132 +96,359 @@ class RA_ENGINE_API Mesh : public Displayable
+        }
+
+      private:
+        AttribArrayDisplayable* m_displayable;
+        int m_idx;
+    };
 
-    ///\todo @dlyr cleanup this mechanism to have something
-    /// extensible. Now the only attribs should be the one defined in
-    /// the enums MeshData, Vec3Data, Vec4Data.
+  protected:
+    std::unique_ptr<globjects::VertexArray> m_vao;
 
-    /// Additionnal vertex vector 3 data handles, stored in Mesh, added
-    std::array<Core::Geometry::TriangleMesh::Vec3AttribHandle, MAX_VEC3> m_v3DataHandle;
-    Core::Geometry::TriangleMesh::Vec3AttribHandle::Container m_dummy3;
-    /// Additionnal vertex vector 4 data handles, stored in Mesh, added
-    std::array<Core::Geometry::TriangleMesh::Vec4AttribHandle, MAX_VEC4> m_v4DataHandle;
-    Core::Geometry::TriangleMesh::Vec4AttribHandle::Container m_dummy4;
+    MeshRenderMode m_renderMode{
+        MeshRenderMode::RM_TRIANGLES}; /// Render mode (GL_TRIANGLES or GL_LINES, etc.)
 
-    // Combined arrays store the flags in this order Mesh, then Vec3 then Vec4 data.
-    // Following the enum declaration above.
-    // Our first VBO index is actually the indices buffer index.
-    // The following are for vertex data.
-    // Each data type has a corresponding openGL attribute number, which is
-    // vbo index - 1 (thus vertex position is VBO number 1 but attribute 0).
+    std::vector<std::unique_ptr<globjects::Buffer>> m_vbos;
+    std::vector<bool> m_dataDirty;
 
-    std::array<uint, MAX_DATA> m_vbos      = {{0}};     /// Indices of our openGL VBOs.
-    std::array<bool, MAX_DATA> m_dataDirty = {{false}}; /// Dirty bits of our vertex data.
+    // Geometry attrib name (std::string) to buffer id (int)
+    // buffer id are indices in m_vbos
+    std::map<std::string, unsigned int> m_handleToBuffer;
 
-    size_t m_numElements{0}; /// number of elements to draw. For triangles this is 3*numTriangles
-                             /// but not for lines.
     /// General dirty bit of the mesh. Must be equivalent of the  "or" of the other dirty flags.
     /// an empty mesh is not dirty
     bool m_isDirty{false};
 };
 
+class VaoIndices
+{
+  protected:
+    std::unique_ptr<globjects::Buffer> m_indices;
+    bool m_indicesDirty{true};
+    /// number of elements to draw. For triangles this is 3*numTriangles
+    /// but not for lines.
+    size_t m_numElements{0};
+};
+
+/// This class handles an attrib array displayable on gpu only, without core
+/// geometry. Use only when you don't need to access the cpu geometry again, or
+/// when you need to specify special indices.
+template <typename I>
+class IndexedAttribArrayDisplayable : public AttribArrayDisplayable, public VaoIndices
+{
+    using IndexType          = I;
+    using IndexContainerType = Ra::Core::AlignedStdVector<IndexType>;
+
+    template <typename T>
+    inline void addAttrib( const std::string& name,
+                           const typename Ra::Core::Utils::Attrib<T>::Container& data ) {
```

Comment:
move the methods definition to the `inl` file

Reply:
Well, I think we need to discuss class name, and then separate class definition to multiples files, and add inl files.